src/oscar/apps/catalogue/migrations/0024_remove_duplicate_attributes.py

```diff
@@ -0,0 +1,65 @@
+# Generated by Django 3.2.9 on 2022-01-25 19:01
+
+from django.db import migrations
+from django.db.models import CharField, Count, Value
+from django.db.models.functions import Concat
+
+def remove_duplicate_attributes(apps, schema_editor):
+    """
+    Removes duplicate attributes that have the same code and product class.
+    """
+    ProductAttribute = apps.get_model('catalogue', 'ProductAttribute')
+    ProductClass = apps.get_model("catalogue", "ProductClass")
+
+    # Instead of iterating over all attributes, we concat the code and product class pk
+    # with a "|" so we can find duplicate attributes in one query.
+    duplicate_attributes = ProductAttribute.objects.annotate(
+        code_and_product_class=Concat('code', Value('|'), 'product_class__pk', output_field=CharField())
+    ).values('code_and_product_class').annotate(
+        same_code_count=Count('code_and_product_class')
+    ).filter(same_code_count__gt=1)
+
+    for attribute in duplicate_attributes:
+        attribute_code, product_class_pk = attribute["code_and_product_class"].split("|")
+        product_class = ProductClass.objects.get(pk=product_class_pk)
+        attributes = ProductAttribute.objects.filter(
+            code=attribute_code,
+            product_class=product_class
+        )
+        used_attributes = attributes.filter(productattributevalue__isnull=False)
+        used_attribute_count = used_attributes.distinct().count()
+
+        # In most cases, the used attributes count will be one or zero as
+        # the dashboard will always show one attribute. If the used attribute
+        # count is one, we exclude this from attributes and remove the others.
+        # If it's zero, we pick the last created and delete others.
```

Comment:
Hi, I don't think you can just pick the last one.
What I think you should do, is use the code that normally picks the attribute.
That means you should use the ProductAttributesContainer to determine which attribute oscar would have chosen and use that as an exclude. The reason is that by default not all database backends will pick the last attribute.

Reply:
I only pick the last one if none of the duplicate attributes have an value relation.
In that case, I don't think it matters which one I keep and which one I delete, as long as there's only one left, right?

If only ONE attribute has a value relation, I keep that one and remove others.

If MORE than ONE attribute has a relation, I try to resolve that as well from line 46 till 53.

Reply:
The way attributes are being added to the product update/create view is as follows:
https://github.com/django-oscar/django-oscar/blob/34c4a1b3448a80fee321e1533d3d93fdd8edd3a8/src/oscar/apps/dashboard/catalogue/forms.py#L274-L285

So every duplicate code, would overwrite the field with that code before (line 282).
So the last on ends up being used inside the dashboard because it's never replaced.

I tried to use this method, but it used .get which raises an error if multiple objects are found.
https://github.com/django-oscar/django-oscar/blob/34c4a1b3448a80fee321e1533d3d93fdd8edd3a8/src/oscar/apps/catalogue/product_attributes.py#L59-L60

So I think I can keep .last() now, as this one is the one being used in dashboard?

Reply:
okok